fast_pair: Add bond management documentation

alstrzebonski opened the pull request:
Jira: NCSDK-29580

NordicBuilder commented:
## CI Information

<sup>_To view the history of this post, clich the 'edited' button above_</sup>
Build number: 2
### Inputs:
#### Sources:
sdk-nrf: PR head: [932426c8479dfe580079426c6b2056e04c5d4b81](https://github.com/nrfconnect/sdk-nrf/pull/17944/commits/932426c8479dfe580079426c6b2056e04c5d4b81)
<details><summary> more details</summary>

sdk-nrf:
> PR head: [932426c8479dfe580079426c6b2056e04c5d4b81](https://github.com/nrfconnect/sdk-nrf/pull/17944/commits/932426c8479dfe580079426c6b2056e04c5d4b81)
> merge base: [194e8fdf26b3c5d20699efed0330b65b45458ae8](https://github.com/nrfconnect/sdk-nrf/commit/194e8fdf26b3c5d20699efed0330b65b45458ae8)
> target head (main): [194e8fdf26b3c5d20699efed0330b65b45458ae8](https://github.com/nrfconnect/sdk-nrf/commit/194e8fdf26b3c5d20699efed0330b65b45458ae8)
> [Diff](https://github.com/nrfconnect/sdk-nrf/compare/194e8fdf26b3c5d20699efed0330b65b45458ae8..932426c8479dfe580079426c6b2056e04c5d4b81)</details>

#### Github labels
| Enabled | Name | Description |
| -- | -- | -- |
|| ci-disabled | _Disable the ci execution_ | 
|| ci-all-test | _Run all of ci, no test spec filtering will be done_ | 
|| ci-force-downstream | _Force execution of downstream even if twister fails_ | 
|| ci-run-twister | _Force run twister_ | 
|| ci-run-zephyr-twister | _Force run zephyr twister_ | 

<details><summary> List of changed files detected by CI (4)</summary>

```
doc
│  ├── nrf
│  │  ├── external_comp
│  │  │  │ bt_fast_pair.rst
│  │  ├── libraries
│  │  │  ├── bluetooth_services
│  │  │  │  ├── services
│  │  │  │  │  │ fast_pair.rst
│  │  ├── releases_and_maturity
│  │  │  ├── releases
│  │  │  │  │ release-notes-changelog.rst
subsys
│  ├── bluetooth
│  │  ├── services
│  │  │  ├── fast_pair
│  │  │  │  │ Kconfig.fast_pair
```
</details>

### Outputs:

#### Toolchain
Version: 
Build docker image: 

#### Test Spec & Results: :white_check_mark: Success; :x: Failure; :orange_circle: Queued; :yellow_circle: Progress; :white_medium_square: Skipped; :warning: Quarantine
- :orange_circle: Toolchain
- :orange_circle: Build twister
- :orange_circle: Integration tests
   - :orange_circle: desktop52_verification
   - :orange_circle: test_ble_nrf_config
   - :orange_circle: test-fw-nrfconnect-ble_samples
   - :orange_circle: test-sdk-find-my

<details><summary>Disabled integration tests</summary>

- 
   - doc-internal
   - test-fw-nrfconnect-apps
   - test-fw-nrfconnect-ble_mesh
   - test-fw-nrfconnect-boot
   - test-fw-nrfconnect-chip
   - test-fw-nrfconnect-fem
   - test-fw-nrfconnect-nfc
   - test-fw-nrfconnect-nrf-iot_cloud
   - test-fw-nrfconnect-nrf-iot_libmodem-nrf
   - test-fw-nrfconnect-nrf-iot_lwm2m
   - test-fw-nrfconnect-nrf-iot_mosh
   - test-fw-nrfconnect-nrf-iot_nrf_provisioning
   - test-fw-nrfconnect-nrf-iot_positioning
   - test-fw-nrfconnect-nrf-iot_samples
   - test-fw-nrfconnect-nrf-iot_serial_lte_modem
   - test-fw-nrfconnect-nrf-iot_thingy91
   - test-fw-nrfconnect-nrf-iot_zephyr_lwm2m
   - test-fw-nrfconnect-nrf_crypto
   - test-fw-nrfconnect-proprietary_esb
   - test-fw-nrfconnect-ps
   - test-fw-nrfconnect-rpc
   - test-fw-nrfconnect-rs
   - test-fw-nrfconnect-tfm
   - test-fw-nrfconnect-thread
   - test-fw-nrfconnect-zigbee
   - test-low-level
   - test-sdk-audio
   - test-sdk-dfu
   - test-sdk-mcuboot
   - test-sdk-pmic-samples
   - test-sdk-sidewalk
   - test-sdk-wifi
   - test-secdom-samples-public
</details>



*Note: This message is automatically posted and updated by the CI*

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-17944
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

peknis commented on doc/nrf/libraries/bluetooth_services/services/fast_pair.rst:
```diff
@@ -47,6 +47,8 @@ With the :kconfig:option:`CONFIG_BT_FAST_PAIR` Kconfig option enabled, the follo
 * :kconfig:option:`CONFIG_BT_FAST_PAIR_CRYPTO_TINYCRYPT`, :kconfig:option:`CONFIG_BT_FAST_PAIR_CRYPTO_OBERON`, and :kconfig:option:`CONFIG_BT_FAST_PAIR_CRYPTO_PSA` - These options are used to select the cryptographic backend for Fast Pair.
   The Oberon backend is used by default.
+* :kconfig:option:`CONFIG_BT_FAST_PAIR_BOND_MANAGER` - The option enables the Fast Pair bond management functionality.
+  See the :ref:`ug_bt_fast_pair_gatt_service_bond_management` for more details.
```
```suggestion
  See :ref:`ug_bt_fast_pair_gatt_service_bond_management` for more details.
```

peknis commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -728,6 +728,10 @@ Bluetooth libraries and services
     * The :kconfig:option:`CONFIG_BT_FAST_PAIR_USE_CASE` Kconfig choice option allowing the user to select their target Fast Pair use case.
       The :kconfig:option:`CONFIG_BT_FAST_PAIR_USE_CASE_UNKNOWN`, :kconfig:option:`CONFIG_BT_FAST_PAIR_USE_CASE_INPUT_DEVICE`, :kconfig:option:`CONFIG_BT_FAST_PAIR_USE_CASE_LOCATOR_TAG` and :kconfig:option:`CONFIG_BT_FAST_PAIR_USE_CASE_MOUSE` Kconfig options represent the supported use cases that can be selected as part of this Kconfig choice option.
+    * The :kconfig:option:`CONFIG_BT_FAST_PAIR_BOND_MANAGER` Kconfig option that enables the Fast Pair bond management functionality.
+      If this option is enabled, the Fast Pair subsystem tracks the Bluetooth bonds created through the Fast Pair Procedure and unpairs them if the Fast Pair Procedure is incomplete or the Account Key associated with the bonds is removed.
```
```suggestion
      If this option is enabled, the Fast Pair subsystem tracks the Bluetooth bonds created through the Fast Pair Procedure and unpairs them if the procedure is incomplete or the Account Key associated with the bonds is removed.
```

annwoj commented on doc/nrf/external_comp/bt_fast_pair.rst:
```diff
@@ -632,6 +632,27 @@ In the Fast Pair not discoverable advertising mode, the Provider informs the lis
+It also unpairs the Fast Pair Bluetooth bonds on Fast Pair factory reset.
+
+Keep in mind that when the Fast Pair subsystem is disabled, the Bluetooth bonds are not tracked.
+Do not perform operations on the Fast Pair Bluetooth bonds apart from the Fast Pair factory reset when the Fast Pair subsystem is disabled.
```
```suggestion
Do not perform operations on Fast Pair Bluetooth bonds, except for the Fast Pair factory reset when the Fast Pair subsystem is disabled.
```

annwoj commented on doc/nrf/external_comp/bt_fast_pair.rst:
```diff
@@ -632,6 +632,27 @@ In the Fast Pair not discoverable advertising mode, the Provider informs the lis
+
+Keep in mind that when the Fast Pair subsystem is disabled, the Bluetooth bonds are not tracked.
+Do not perform operations on the Fast Pair Bluetooth bonds apart from the Fast Pair factory reset when the Fast Pair subsystem is disabled.
+Also do not perform Bluetooth bonding when the Fast Pair subsystem is disabled and bond management functionality is enabled as it may lead to those new bonds being incorrectly considered as Fast Pair bonds by the Fast Pair subsystem.
```
```suggestion
Additionally, do not perform Bluetooth bonding when the Fast Pair subsystem is disabled and bond management functionality is enabled, as this could cause the subsystem to incorrectly identify new bonds as Fast Pair bonds.
```

annwoj commented on subsys/bluetooth/services/fast_pair/Kconfig.fast_pair:
```diff
@@ -183,10 +183,12 @@ config BT_FAST_PAIR_BOND_MANAGER
-	  information. You can enable this option during DFU, but you cannot disable it during DFU
-	  if you used it earlier unless you erase whole Fast Pair non-volatile storage. Otherwise,
-	  the Fast Pair enable operation will fail.
+	  is disabled. Also do not perform Bluetooth bonding when the Fast Pair subsystem is
+	  disabled as it may lead to those new bonds being incorrectly considered as Fast Pair bonds
```
```suggestion
	  is disabled. Additionally, do not perform Bluetooth bonding when the Fast Pair subsystem is
	  disabled as it could cause the subsystem to incorrectly identify new bonds as Fast Pair bonds.
```

annwoj commented on subsys/bluetooth/services/fast_pair/Kconfig.fast_pair:
```diff
@@ -183,10 +183,12 @@ config BT_FAST_PAIR_BOND_MANAGER
-	  the Fast Pair enable operation will fail.
+	  is disabled. Also do not perform Bluetooth bonding when the Fast Pair subsystem is
+	  disabled as it may lead to those new bonds being incorrectly considered as Fast Pair bonds
+	  by the Fast Pair subsystem. When this option is enabled, non-volatile storage is used to
```
```suggestion
	  When this option is enabled, non-volatile storage is used to
```

annwoj commented on subsys/bluetooth/services/fast_pair/Kconfig.fast_pair:
```diff
@@ -183,10 +183,12 @@ config BT_FAST_PAIR_BOND_MANAGER
+	  is disabled. Also do not perform Bluetooth bonding when the Fast Pair subsystem is
+	  disabled as it may lead to those new bonds being incorrectly considered as Fast Pair bonds
+	  by the Fast Pair subsystem. When this option is enabled, non-volatile storage is used to
+	  store the bond information. You can enable this option during DFU, but you cannot disable
+	  it during DFU if you used it earlier unless you erase whole Fast Pair non-volatile
```
This sentence is also in the line 649 of bt_fast_pair (included Pekka's suggestion). Please decide which version works best

```suggestion
	  store the bond information. You can enable the functionality during DFU, but if you used it earlier, 
	  you cannot disable it during DFU unless you erase the entire Fast Pair non-volatile
```

MarekPieta commented on doc/nrf/external_comp/bt_fast_pair.rst:
```diff
@@ -632,6 +632,27 @@ In the Fast Pair not discoverable advertising mode, the Provider informs the lis
+Otherwise, the Fast Pair enable operation will fail.
+
+By default, the Fast Pair bond management functionality is disabled.
+Some use cases of the Google Fast Pair application require the functionality to be enabled.
```
Please mention that use-cases that need the feature imply it.

alstrzebonski commented:
Addressed review comments. Removed some of the duplicated details form `bt_fast_pair.rst` and left them only in Kconfig help which is referred in the `.rst` file. Please rereview.

annwoj commented on doc/nrf/external_comp/bt_fast_pair.rst:
```diff
@@ -632,6 +632,24 @@ In the Fast Pair not discoverable advertising mode, the Provider informs the lis
+See the :kconfig:option:`CONFIG_BT_FAST_PAIR_BOND_MANAGER` Kconfig option help for more datails about using the functionality.
+
+The Fast Pair bond management functionality is disabled by default.
+Following use cases of the Google Fast Pair application require it to be enabled:
```
```suggestion
You must enable it for the following use cases of the Google Fast Pair application:
```

alstrzebonski replied:
Changed to `It is highly recommended to enable it for the following use cases of the Google Fast Pair application:` as must was too strong.

kapi-no commented on doc/nrf/external_comp/bt_fast_pair.rst:
```diff
@@ -632,6 +632,24 @@ In the Fast Pair not discoverable advertising mode, the Provider informs the lis
+See the :kconfig:option:`CONFIG_BT_FAST_PAIR_BOND_MANAGER` Kconfig option help for more datails about using the functionality.
+
+The Fast Pair bond management functionality is disabled by default.
+Following use cases of the Google Fast Pair application require it to be enabled:
```
Are you sure about this claim?

MarekPieta replied:
Consider using e.g. `strongly suggest` term

alstrzebonski commented:
Adressed one last review comment. Please rereview.

alstrzebonski commented:
Rebased on main.

MarekPieta commented on doc/nrf/external_comp/bt_fast_pair.rst:
```diff
@@ -632,6 +632,24 @@ In the Fast Pair not discoverable advertising mode, the Provider informs the lis
+To enable the Fast Pair bond management functionality, use the :kconfig:option:`CONFIG_BT_FAST_PAIR_BOND_MANAGER` Kconfig option.
+When this functionality is enabled, the Fast Pair subsystem tracks the Bluetooth bonds created through the Fast Pair Procedure and unpairs them if the procedure is incomplete or the Account Key associated with the bonds is removed.
+It also unpairs the Fast Pair Bluetooth bonds on Fast Pair factory reset, because the factory reset removes all Account Keys stored on device.
+See the :kconfig:option:`CONFIG_BT_FAST_PAIR_BOND_MANAGER` Kconfig option help for more datails about using the functionality.
```
Can we mention that enabling the functionality imposes some additional limitations (Fast Pair needs to be enabled while performing operations on Bluetooth bonds)? I would suggest to mention that:
`Enabling the feature imposes additional limitations related to enabling Fast Pair.`

MarekPieta replied:
The option is enabled by default for some of the use-cases. I am worried that users might overlook the limitation because of that.

MarekPieta commented on doc/nrf/external_comp/bt_fast_pair.rst:
```diff
@@ -632,6 +632,24 @@ In the Fast Pair not discoverable advertising mode, the Provider informs the lis
+See the :kconfig:option:`CONFIG_BT_FAST_PAIR_BOND_MANAGER` Kconfig option help for more datails about using the functionality.
+
+The Fast Pair bond management functionality is disabled by default.
+It is highly recommended to enable it for the following use cases of the Google Fast Pair application:
```
The listed use-cases imply bond management. Can we rephrase a bit to clearly indicate that (current wording suggests that user needs to enable it manually)?

alstrzebonski replied:
There's a difference between the use case and the kconfig that enables the use case. User should see the kconfig to find the required feature set.

alstrzebonski replied:
From the use case section: `The selected Kconfig option configures the Fast Pair features and extensions to satisfy the `Fast Pair Device Feature Requirements`_ for your target use case.`

alstrzebonski replied:
I'd rather leave it as is. I don't want to introduce the use case Kconfigs at this point of documentation. They are introduced later.

MarekPieta replied:
Discussed offline. We will use `Make sure` syntax here.

MarekPieta commented on subsys/bluetooth/services/fast_pair/Kconfig.fast_pair:
```diff
@@ -177,16 +177,22 @@ config BT_FAST_PAIR_BOND_MANAGER
+	  functionality is enabled, as this could cause the subsystem to incorrectly identify new
+	  bonds as Fast Pair bonds. To conclude, the easiest and safest option is to keep the Fast
+	  Pair enabled while using Bluetooth to ensure proper behavior. With this functionality, the
+	  non-volatile storage is used to store the bond information. You can enable the
```
Can we make the DFU part as a separate section to improve readability? Same for part starting with `When the Fast Pair subsystem is disabled...`

MarekPieta commented on doc/nrf/external_comp/bt_fast_pair.rst:
```diff
@@ -632,6 +632,25 @@ In the Fast Pair not discoverable advertising mode, the Provider informs the lis
+To enable the Fast Pair bond management functionality, use the :kconfig:option:`CONFIG_BT_FAST_PAIR_BOND_MANAGER` Kconfig option.
+When this functionality is enabled, the Fast Pair subsystem tracks the Bluetooth bonds created through the Fast Pair Procedure and unpairs them if the procedure is incomplete or the Account Key associated with the bonds is removed.
+It also unpairs the Fast Pair Bluetooth bonds on Fast Pair factory reset, because the factory reset removes all Account Keys stored on device.
+Enabling the functionality imposes additional limitations related to enabling Fast Pair.
```
`... enabling Fast Pair in runtime.`
